### PR TITLE
Implemented peephole optimizations

### PR DESCRIPTION
**Pull request by Stoeoef**
This time with a more "static matching" approach. I'll add some more tests and maybe some good patterns.

### CODE REVIEW

File: src/main/scala/mjis/AssemblerFileGenerator.scala [hunk 1]
```diff
@@ -92,8 +92,8 @@ class MjisAssemblerFileGenerator(input: AsmProgram, config: Config) extends Asse
         if (function.isLoopHeader(block)) emit(".p2align 4,,15")
         emit(
           (block match {
-            case function.prologue => "# Prologue"
-            case function.epilogue => "# Epilogue"
+            case x if x == function.prologue => "# Prologue"
```
Comment (Kha) [inline src/main/scala/mjis/AssemblerFileGenerator.scala]: :question:

File: src/main/scala/mjis/PeepholeOptimizer.scala [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import scala.collection.mutable
+import mjis.asm._
+import scala.collection.mutable.AbstractBuffer
+import scala.language.existentials
+import scala.collection.mutable.ListBuffer
+
+class PeepholeOptimizer(input: AsmProgram) extends Phase[AsmProgram] {
+  var num_replacements = 0
+
+  override def findings: List[mjis.Finding] = List()
```
Comment (Kha) [inline src/main/scala/mjis/PeepholeOptimizer.scala]: Now implemented in `Phase`

File: src/main/scala/mjis/PeepholeOptimizer.scala [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  def getResult(): AsmProgram = {
+    optimize(input)
+  }
+
+  def optimize(input: AsmProgram): AsmProgram = {
+    var result = new AsmProgram
```
Comment (Kha) [inline src/main/scala/mjis/PeepholeOptimizer.scala]: Just modify the input in-place

File: src/main/scala/mjis/PeepholeOptimizer.scala [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    result.successors ++= input.successors
+    result.instructions ++= optimizeSequence(input.instructions)
+    result
+  }
+
+  def optimizeSequence(instructions: ListBuffer[Instruction]): ListBuffer[Instruction] = {
+    var newInstructions = ListBuffer[Instruction]()
+    var oldInstructions = instructions
+    var changed = false
+    // Repeat "peepholeing" until one pass doesn't change any instruction
```
Comment (Kha) [inline src/main/scala/mjis/PeepholeOptimizer.scala]: Shouldn't it be enough to go back 3 instructions on a change?

**Comment by Stoeoef** [comment]
Updated. Unfortunately, more than 90% of all optimized instructions are "mov x x" instructions, which have already been optimized by `AssemblerFileGenerator` . So, the actual instruction count reduction is just about 1.3%

**Comment by jspam** [comment]
The flatMap in `AssemblerFileGenerator` can be removed, it was just a temporary replacement until we had the real thing.

**Comment by Stoeoef** [comment]
Done.

**Comment by Kha** [comment]
Suggestions for further patterns that may actually be beneficial:
* `mov $0, r` => `xor r, r`
* `cmpl` => `test` (where applicable)

**Comment by Stoeoef** [comment]
Updated to latest master.
Thanks for the patterns, I'll see if I can implement them.

**Comment by TehMillhouse** [comment]
Something went wrong when you updated to the current master, making it impossible to simply merge this. I'll fix that and open a new PR with only your commit. (Don't worry, the commit will still show as yours)

**Pull request closed without merging**